[cmd/opampsupervisor] Add a healthcheck endpoint for the Supervisor

douglascamata opened the pull request:
#### Description

This PR adds a healthcheck endpoint to the Supervisor. It runs in its own dedicated HTTP server. For now the servers uses port `23233` (inspired on the Collector's default healthcheck port: `13133`). The Supervisor's healthcheck port is not configurable at the moment, but it could be in the future.

#### Link to tracking issue
Fixes #40529.

#### Testing

Unit test added.

douglascamata committed:
[cmd/opampsupervisor] Implement a healthcheck endpoint for the Supervisor

douglascamata committed:
Use fixed port for the healthcheck server

douglascamata committed:
Add changelog entry

douglascamata committed:
Fix some linter warnings

douglascamata committed:
Use listener to ensure server can listen on healthcheck port

douglascamata committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector-contrib into supervisor-healthcheck-endpoint

douglascamata committed:
Make linter happy

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Make linter happy again

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

atoulme commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -689,6 +697,52 @@ func (s *Supervisor) startOpAMPClient() error {
+
+	healthCheckServerPort := 23233
+	s.healthCheckServer = &http.Server{
+		Addr:        fmt.Sprintf(":%d", healthCheckServerPort),
```
this is binding to all network interfaces

douglascamata committed:
Make healthcheck server listen on localhost only

douglascamata committed:
Make healthcheck port configurable

By default, it's zero and in this case we don't start a healthcheck
server.

atoulme commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -689,6 +697,56 @@ func (s *Supervisor) startOpAMPClient() error {
+
+	healthCheckServerPort := s.config.HealthCheck.Port
+	s.healthCheckServer = &http.Server{
+		Addr:        fmt.Sprintf("localhost:%d", healthCheckServerPort),
```
I'm sorry. That's better, but you want this configurable too - because in docker or ECS type envs, you'd like to bind to 0.0.0.0 or to a specific network interface.

douglascamata replied:
No worries, I can make the host interface configurable and default to `localhost`. Should be a small change and I can take some time to rebase with `main` because I'm almost 100 commits behind.

atoulme reviewed (approved):
this can land as is imo - I have a nit about the network interface but I'm ok if we address it separately based on more needs being shored up

evan-bradley commented on cmd/opampsupervisor/supervisor/config/config.go:
```diff
@@ -266,6 +271,17 @@ type Telemetry struct {
 }
 
+type HealthCheck struct {
+	Port int `mapstructure:"port"`
```
Sorry, probably should have taken a closer look and called this out earlier: it may be overkill, but I think we should use [confighttp.ServerConfig](https://pkg.go.dev/go.opentelemetry.io/collector/config/confighttp#ServerConfig) here. It keeps us consistent with Collector receivers, the healthcheck(v2) extension, and offers all the options we could need out of the box.

douglascamata replied:
Working on it.

douglascamata committed:
Configure healthcheck server with `httpserver.ServerConfig`

douglascamata committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector-contrib into supervisor-healthcheck-endpoint

douglascamata commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -696,6 +704,61 @@ func (s *Supervisor) startOpAMPClient() error {
+	healthCheckServerPort := s.config.HealthCheck.Port()
+	server, err := s.config.HealthCheck.ToServer(
+		context.Background(),
+		nil,
```
@evan-bradley I wasn't about what to put here in the `component.Host` field of this method call. Do you have suggestions? 🤔

evan-bradley replied:
Hm, that is a little tricky. I think we can just make our own no-op host since the interface is fairly simple, you can see it here: https://pkg.go.dev/go.opentelemetry.io/collector/component#Host. Returning a `nil` map for `GetExtensions` should be sufficient for the method implementation.

douglascamata committed:
Tidy up

douglascamata committed:
Update changelog entry

douglascamata committed:
Create a nopHost for the healthcheck server

atoulme reviewed (approved):
LGTM - @evan-bradley please take one more look and feel free to merge, thanks!

evan-bradley reviewed (commented):
Overall looks good, just a few suggestions. Two additional requests if you don't mind:

1. Can we add this config option to one of the YAML files in `testdata/supervisor` and verify the config is parsed correctly? This would probably be an E2E test.
2. Can you document this option in the readme?

evan-bradley commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -696,6 +704,67 @@ func (s *Supervisor) startOpAMPClient() error {
+
+	return nil
+}
+
+type nopHost struct{}
```
That's a fair point. I'm alright leaving it out then, thanks for the explanation.

douglascamata committed:
Add a healthcheck section to the README

douglascamata committed:
Make healthcheck failures return 503s

douglascamata committed:
Ensure `nopHost` implements `component.Host`

douglascamata committed:
Implement e2e test for healthcheck server and fix it

douglascamata committed:
Fix mistake in the README

douglascamata commented:
@evan-bradley all the feedback taken care of, sir!

douglascamata commented:
Oh wait, got some weird test failures.

douglascamata committed:
Fix tests

douglascamata commented:
@evan-bradley now the tests should all pass.

evan-bradley commented:
It looks like there's a leaking goroutine in one of the E2E tests, I assume the new one unless something changed on main. I took a quick look and nothing jumped out at me.

douglascamata committed:
Fix go routine leak in `newUnstartedOpAMPServer`

The interactions of the callbacks with the `connectedChan` had the risk
to introduce a go-routine leak in the tests if
`waitForSupervisorConnection` was not called before the Supervisor was
shutdown.

Closing the `connectedChan` in the OpAMP server shutdown function
allowed me to detect this.

douglascamata committed:
Wait for Supervisor connection in happy-path healthcheck test

douglascamata committed:
Add a healthcheck test where the opamp backend is unreachable

douglascamata commented on cmd/opampsupervisor/e2e_test.go:
```diff
@@ -2256,6 +2256,47 @@ func TestSupervisorOpAmpServerPort(t *testing.T) {
 }
 
+func TestSupervisorHealthCheckServer(t *testing.T) {
+	server := newOpAMPServer(
+		t,
+		defaultConnectingHandler,
+		types.ConnectionCallbacks{},
+	)
```
I found and fixed the test Go routine leak in [71b0c5b](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41565/commits/71b0c5b2e0243ed964f5bb4d0e694e582d351ee6). It was a tricky one! The commit's description contains more details about it, if you are interested. What do you think?

The new test case requested was added in [2388c08](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41565/commits/2388c0881e6f3a74fdd9a9fa5dd72137977c831c).

Thanks for the review, @evan-bradley.

evan-bradley replied:
Nice work! That was a really subtle bug, thanks for the detailed explanation in the commit message.

evan-bradley reviewed (approved):
Thanks for being so responsive and diligent throughout the review process, this one turned out to have a few more considerations than I was expecting at the onset.
